app/services/katello/registration_manager.rb

```diff
@@ -0,0 +1,156 @@
+module Katello
+  class RegistrationManager
+    class << self
+      private :new
+
+      def unregister_host(host, options = {})
+        organization_destroy = options.fetch(:organization_destroy, false)
+        unregistering = options.fetch(:unregistering, false)
+
+        if !organization_destroy && host.subscription_facet.try(:uuid)
+          ::Katello::Resources::Candlepin::Consumer.destroy(host.subscription_facet.uuid)
+        end
+
+        begin
+          ::Katello.pulp_server.extensions.consumer.delete(host.content_facet.uuid) if host.content_facet.try(:uuid)
+        rescue RestClient::ResourceNotFound
+          Rails.logger.error(_("Pulp Consumer %s has already been removed") % host.content_facet.uuid)
+        end
+
+        host.subscription_facet.try(:destroy!)
+
+        if unregistering
+          remove_host_artifacts(host)
+        elsif organization_destroy
+          host.content_facet.try(:destroy!)
+          destroy_host_artifacts(host)
+        else
+          host.content_facet.try(:destroy!)
+          unless host.destroy
+            fail host.errors.full_messages.join('; ')
+          end
+        end
+      end
+
+      def register_host(host, consumer_params, content_view_environment, activation_keys = [])
+        unless host.new_record?
+          host.save!
+          unregister_host(host, :unregistering => true)
+          host.reload
+        end
+
+        unless activation_keys.empty?
+          content_view_environment ||= lookup_content_view_environment(activation_keys)
+          set_host_collections(host, activation_keys)
+        end
+        fail _('Content View and Environment not set for registration.') if content_view_environment.nil?
+
+        host.save!
+        host.content_facet = plan_content_facet(host, content_view_environment)
+        host.subscription_facet = plan_subscription_facet(host, activation_keys, consumer_params)
+        host.save!
+
+        cp_create = ::Katello::Resources::Candlepin::Consumer.create(content_view_environment.cp_id, consumer_params, activation_keys.map(&:cp_name))
```

Comment:
One thing to consider on both the register and unregister methods,  pretend that any one of these backend services throws an ISE or 400, what happens in that case?  and what data is left over?  Is that situation okay?  can we clean it up in any way?  

I would consider three scenarios:

Candlepin throws an ISE/400 on register
Pulp throws an ISE/400 on register
host.save!  fails spectacularly   (It looks like maybe this is handled a bit)

The dynflow tasks were good (or bad) at handling this depending on how you looked at it, but you generally could ensure the consistency by resuming the task (assuming the error was fixable outside of the task).   Attempting to ensure consistency here would be good